pinot-controller/src/main/java/org/apache/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1029,6 +1039,47 @@ void updateInstanceStatesForNewConsumingSegment(Map<String, Map<String, String>>
     }
   }
 
+  /**
+   * Handles segment movement between instances.
+   * If the new consuming segment is served by a different set of servers than the committed segment, notify the
+   * servers no longer serving the stream partition to remove the ingestion metrics. This can prevent servers from
+   * emitting high ingestion delay alerts on stream partitions no longer served.
+   */
+  private void handleSegmentMovement(String realtimeTableName, Map<String, Map<String, String>> instanceStatesMap,
+      String committedSegment, String newConsumingSegment) {
+    Set<String> oldInstances = instanceStatesMap.get(committedSegment).keySet();
+    Set<String> newInstances = instanceStatesMap.get(newConsumingSegment).keySet();
+    if (newInstances.containsAll(oldInstances)) {
+      return;
+    }
+    Set<String> instancesNoLongerServe = new HashSet<>(oldInstances);
+    instancesNoLongerServe.removeAll(newInstances);
+    LOGGER.info("Segment movement detected for committed segment: {} (served by: {}), "
+            + "consuming segment: {} (served by: {}) in table: {}, "
+            + "sending message to instances: {} to remove ingestion metrics", committedSegment, oldInstances,
+        newConsumingSegment, newInstances, realtimeTableName, instancesNoLongerServe);
+
+    ClusterMessagingService messagingService = _helixManager.getMessagingService();
+    List<String> instancesSent = new ArrayList<>(instancesNoLongerServe.size());
+    for (String instance : instancesNoLongerServe) {
+      Criteria recipientCriteria = new Criteria();
+      recipientCriteria.setInstanceName(instance);
+      recipientCriteria.setRecipientInstanceType(InstanceType.PARTICIPANT);
+      recipientCriteria.setResource(realtimeTableName);
+      recipientCriteria.setPartition(committedSegment);
+      recipientCriteria.setSessionSpecific(true);
+      IngestionMetricsRemoveMessage message = new IngestionMetricsRemoveMessage();
```

Comment:
I feel that we should rename `IngestionMetricsRemoveMessage` to something else so that it can be reused for other purposes as well in future.  

e.g. `RemoveInstanceMessage`

Reply:
I prefer the current naming convention. Creating a custom message is not costly, and having a separate message for each specific task is more desirable. This approach adheres to the ‘single responsibility’ principle.